logic/src/commonMain/kotlin/com/wire/kalium/logic/data/message/Message.kt

```diff
@@ -60,6 +61,42 @@ sealed interface Message {
         val visibility: Visibility
     }
 
+    data class Ephemeral(
+        val expireAfterMillis: Long,
+        val selfDeletionStartDate: Long?,
+        val ephemeralMessage: Regular
+    ) : Standalone {
+
+        fun actualExpireAfterMillis(): Long {
+            return if (selfDeletionStartDate == null) {
+                expireAfterMillis
+            } else {
+                // when we ask for expire after millis after the self deletion start data is determined
+                // we calculate it by knowing the time difference
+                expireAfterMillis - (Clock.System.now().toEpochMilliseconds() - selfDeletionStartDate)
+            }
+        }
+
+        fun isDeletionOngoing(): Boolean {
+            return selfDeletionStartDate != null
+        }
+
+        override val id: String
+            get() = ephemeralMessage.id
+        override val content: MessageContent.Regular
+            get() = ephemeralMessage.content
+        override val conversationId: ConversationId
+            get() = ephemeralMessage.conversationId
+        override val date: String
+            get() = ephemeralMessage.date
+        override val senderUserId: UserId
+            get() = ephemeralMessage.senderUserId
+        override val status: Status
+            get() = ephemeralMessage.status
+        override val visibility: Visibility
+            get() = ephemeralMessage.visibility
+    }
```

Comment:
Why would Ephemeral be its own type?

It makes it so we need to FIRST know if it's ephemeral and THEN check what's the type/content.

Initially, I don't see any need for having a new "top-level" message type at all.


What if instead a nullable `EphemeralData` is added to `Regular` messages?
This could contain `expireAfterMillis` and `selfDeletionStartDate`.


So we can: check if it's a text/asset/whatever, and THEN check if the `ephemeralData` is null or not.

Reply:
Agreed good point with well explained reason why u think this way I am convinced :)